src/extension/monitor/core/src/test/java/org/geoserver/monitor/auditlog/AuditLoggerTest.java

```diff
@@ -0,0 +1,103 @@
+/* (c) 2019 Open Source Geospatial Foundation - all rights reserved
+ * This code is licensed under the GPL 2.0 license, available at the root
+ * application directory.
+ */
+package org.geoserver.monitor.auditlog;
+
+import static org.hamcrest.Matchers.arrayWithSize;
+import static org.hamcrest.Matchers.emptyArray;
+import static org.hamcrest.Matchers.greaterThan;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertThat;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import org.geoserver.monitor.MonitorConfig;
+import org.geoserver.monitor.RequestData;
+import org.geoserver.platform.GeoServerExtensions;
+import org.geoserver.platform.GeoServerResourceLoader;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.springframework.context.event.ContextClosedEvent;
+import org.springframework.context.support.GenericApplicationContext;
+
+public class AuditLoggerTest {
+
+    @Rule public TemporaryFolder directory = new TemporaryFolder();
+
+    private File bad;
+    private File good;
+    private GeoServerResourceLoader loader;
+    private MonitorConfig config;
+    private GenericApplicationContext context;
+
+    @Before
+    public void setUp() throws Exception {
+        this.bad = this.directory.newFolder("bad");
```

Comment:
I don't understand this one... even if two threads end up performing the initialization, won't they both stop when they [find the system variable](https://github.com/geoserver/geoserver/blob/a38bc435cb9889f60a84fb67c815f5ad6c218035/src/extension/monitor/core/src/main/java/org/geoserver/monitor/auditlog/AuditLogger.java#L86) and end up ignoring the "bad" config value? The config field is used only if the path retrieved from the system property is null or empty, I don't see code resetting it to null or empty.

I would expect the race condition to have two dumpers created against the same directory and end up overwriting the first few request worth of logs, likely causing the log files to be invalid.

Reply:
No feedback on this comment?

Reply:
There are three possible bad effects of this race condition that I've seen.

1. What you mentioned where multiple RequestDumper threads will be created that try to write to the same log file.  Only one thread will succeed so a few requests will not be logged and there may be a few extra RequestDumper threads that will never exit.  This is not dependent on the system property and is self-correcting other than a minor leak.
2. Multiple RequestDumper threads will be created that try to write to different log files.  A few audit log files may be created with only a single request in them and there may be a few extra RequestDumper threads that will never exit.  This is not dependent on the system property and is self-correcting other than a minor leak.
3. The most severe bug is dependent on the system property and will not automatically correct itself.  The only viable option to fix this with a clustered GeoServer environment using a shared data directory is to restart any bugged GeoServer nodes until they are in a good state.  initDumper() sets the AuditLogger path to the system property, creates the resource from the path, and then sets the path to the property from the config file.  When multiple threads call initDumper() at the same time, it is possible for one thread to set the path to the config file value between the other thread setting it to the system property and creating the resource.

Reply:
My unit test checks for all of these possible error conditions.